distributed/worker.py

```diff
@@ -1452,19 +1462,30 @@ def add_task(
                         "Asked to compute pre-existing result: %s: %s", key, ts.state
                     )
                     self.send_task_state_to_scheduler(ts)
+                    # FIXME: Can we pull this off without a return to ensure
+                    # that we do not miss anything following?
                     return
                 if ts.state in IN_PLAY:
-                    return
+                    # XXX I believe we only require this special treatment in
+                    # case we're already executing or in memory to avoid
+                    # resetting / aborting the computation. Then again, if the
+                    # execution is ongoing, finishing it _should_ transfer it to
+                    # memory regardless of the start state, shouldn't it?
+
+                    # State waiting is ambiguous since it is used for
+                    # dependencies which are waiting to be fetched (no runspec)
+                    # and for tasks ready to be computed (with runspec)
+                    if ts.state == "waiting":
+                        ts.runspec = runspec
                 if ts.state == "erred":
                     ts.exception = None
                     ts.traceback = None
                 else:
-                    ts.state = "waiting"
+                    self.transition(ts, "waiting", runspec=runspec)
```

Comment:
Not setting the runspec for these cases can actually lead to a deadlock. I've been hunting this deadlock for a while now and I am very confident this is the root cause for #4133

I also have a suspicion that this is causing #4173 but I cannot reproduce locally

Fixes are still incomplete and I will need more time fixing CI since I cannot reproduce everything. I'm a bit confused about what the "waiting" state is actually supposed to flag since this is slightly different for dependencies and runnable tasks (see in code comment). Any pointers and opinions appreciated